### PR TITLE
Add sys_heap heap allocator

### PR DESCRIPTION
**Pull request by andyross**
This has been sitting around without much controversy.  Let's resurrect it and get it into the tree.  Note that this is just the bare heap implementation under a new API.  I'll push the k_mem_pool wrapper separately (which, while it can implement the full API, can't really be a 100% compatible replacement because of the way the memory layout of the original pool was documented).  It can live via a kconfig switch for a release before deprecating the original.

Note also that this doesn't include @npitre's optimization series, since I don't know if he has time to work on that at the moment.  Those too can go in on top of this once merged.

### CODE REVIEW

**Comment by npitre** [comment]
On Mon, 30 Mar 2020, Andy Ross wrote:

> Note also that this doesn't include @npitre&#39;s optimization series, since I don't know if he has time to work on that at the moment.  Those too can go in on top of this once merged.

I should be able to find some time to look at it, and port my changes to it. 
Time is plentiful these days.

**Review by npitre — APPROVED** [review]

**Comment by andyross** [comment]
The development work caught up with the older PR, so now this contains the k_mem_pool compatibility layer, selectable via CONFIG_MEM_POOL_HEAP_BACKEND.  It's implemented on top of a very thin "k_heap" layer, which itself does nothing but add synchronization to the sys_heap underlayment.

Please re-review the new stuff if you looked at this earlier.

**Comment by npitre** [comment]
lots of test failure with native_posix and native_posix_64.
Is that expected?

**Comment by andyross** [comment]
Fix whackamole regressions (see last patch in the series) -- I expanded the default size heuristic for the mem_pool backend at the last minute and broke some tests that actually relied on small heaps filling up exactly as they did before.  The new allocator... failed to fail.

@npitre no, it's for sure all supposed to be passing.

**Comment by andyross** [comment]
Hm... that timeout in CI seems spurious.  It looks like the emulator hung, maybe?  (This is the riscv "renode" platform, not qemu).   I'm no renode expert, but I can't reproduce locally.  It seems to work fine, though surprisingly slowly (it takes 20-30 seconds to finish on my 5 GHz box!).

**Comment by andrewboie** [comment]
looks like from the log there are some build failures on x86_64. Maybe need a UL suffix on some constants?
The log is huge, to the point where shippable couldn't parse it, I only looked briefly

```
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/tests/lib/heap/src/main.c:27:22: error: integer overflow in expression of type 'int' results in '-2147483648' [-Werror=overflow]
   27 | # define MEMSZ (1024 * CONFIG_SRAM_SIZE)
      |                      ^
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/include/sys/util.h:110:28: note: in definition of macro 'MIN'
  110 | #define MIN(a, b) (((a) < (b)) ? (a) : (b))
      |                            ^
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/tests/lib/heap/src/main.c:30:37: note: in expansion of macro 'MEMSZ'
   30 | #define BIG_HEAP_SZ MIN(256 * 1024, MEMSZ / 3)
      |                                     ^~~~~
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/tests/lib/heap/src/main.c:37:15: note: in expansion of macro 'BIG_HEAP_SZ'
   37 | void *heapmem[BIG_HEAP_SZ / sizeof(void *)];
      |               ^~~~~~~~~~~
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/tests/lib/heap/src/main.c:27:22: error: integer overflow in expression of type 'int' results in '-2147483648' [-Werror=overflow]
   27 | # define MEMSZ (1024 * CONFIG_SRAM_SIZE)
```

**Comment by andyross** [comment]
I have genuinely no idea what that is.  This test builds just fine in the PR.   CONFIG_SRAM_SIZE is a value in kb, and you can only overflow that in 32 bits if we expose over 4G in memory space, which we clearly don't.  Is this some variant config builder it's trying out and ignoring or something?  I mean, yeah, that kconfig value should be cast to a size_t for this particular line, but that's definitely not an error we see in practice.

But regardless, that's not flagged as an error by CI anyway.  What seems to be happening at the bottom of the log is that the renode process is stuck, or not being killed.  @nashif : I'm guessing this test is actually timing out (renode seems really slow, and it's a CPU-bound test) and maybe the "kill the test" logic in sanitycheck isn't working for that simulator?

**Comment by andyross** [comment]
Rebase and add a longer timeout to the lib/heap test, guessing that's what the issue with CI is.

**Comment by andyross** [comment]
Hm.... looks like CI is stuck on that same spot.  @nashif, when you get a chance if you could look at this and see if you can divine what the issue is?  Maybe it's the version of renode?  (I'm on the 1.9.0 fedora package).

I might as well add the size_t cast while I'm at it, though those errors in the log don't reproduce in-tree, so I'm not sure where it's coming from.

**Review by tyrantlizard77 — COMMENTED** [review]
Two questions, both related to use cases where k_heap is used to allocate a thread stack.

1. How hard would it be, as a future extension, to support allocating blocks of memory whose base pointer is aligned to some runtime provided value? It's currently hard to allocate thread stacks on platforms where the stack object must be aligned to its own size.

2. On platforms where this is not the case, such as x86, the stacks still must be page-aligned (or at least aligned to some fixed value known at build time). If I create a heap whose buffer address is page aligned, and I only allocate page-sized blocks of memory, are all such allocations guaranteed to also be page-aligned?

**Comment by andrewboie** [comment]
One other comment: if k_heap is a complete replacement for k_mem_pool (and it seems like it is), I'd like to see all the k_mem_pool APIs __deprecated so they can ultimately be removed in a couple releases, rather than having to maintain on an ongoing basis both sets. If you don't want to do that in this PR please at least file a GH issue to make this happen.

**Comment by npitre** [comment]
On Wed, 8 Apr 2020, Andrew Boie wrote:

> @andrewboie commented on this pull request.
> 
> 1. How hard would it be, as a future extension, to support allocating blocks of memory whose base pointer is aligned to some provided value? It's currently hard to allocate thread stacks on platforms where the stack object must be aligned to its own size.

Moderately easy but memory fragmentation will occur.

> 2. On platforms where this is not the case, such as x86, the stacks still must be page-aligned. If I create a heap whose buffer address is page aligned, and I only allocate page-sized blocks of memory, are all such allocations guaranteed to also be page-aligned?

No. There is some metadata around memory blocks that is sure to misalign 
things. And assuming (1) is implemented then you'll waste every other 
page.

For those things I think it is way way better to use a mem_slab instead.

**Review by tyrantlizard77 — APPROVED** [review]
> Moderately easy but memory fragmentation will occur.

That's probably an acceptable trade-off, and generally understood as a known limitation when you don't have virtual memory.

> No. There is some metadata around memory blocks that is sure to misalign things. And assuming (1) is implemented then you'll waste every other page. For those things I think it is way way better to use a mem_slab instead.

mem_slab of page-sized blocks is only good if I just need 1 page at a time. If I want a stack of 3 pages of contiguous memory for a 12K stack buffer, a slab doesn't work (unless of course all slabs are 12K)

At any rate, it looks like we can add aligned allocation later if we feel we need to, and we've certainly gotten this far as an OS without supporting heap-allocated stacks, so +1 from me.

**Comment by andyross** [comment]
@npitre beat me to the answer on alignment.  It's easy enough for the "only occasional allocations are aligned" case -- you just grab a larger chunk, slice off the misaligned prefix and return that to the heap as a smaller bit, basically paying the cost in some extra fragmentation.  Maybe that should just be added to this PR, though I need to work on some xtensa stuff more pressingly.

It's definitely not a good choice for page-sized allocation, we'd want a different allocator for that (or.. I dunno, actually you could use the same code, but make chunks 4096 bytes instead of 8, and jigger the field successors so the metadata gets stored elsewhere -- would be some major surgery but would work well enough).

As far as deprecating k_mem_pool_*(), I think that makes sense.  But note that we have some auxilliary APIs (mbox, msgq, et. al.) which are written to assume the same k_mem_block struct that we have from mem_pool, so we can't completely deprecate it.

**Comment by andrewboie** [comment]
> Maybe that should just be added to this PR, though I need to work on some xtensa stuff more pressingly.

Nah, no rush.

> It's definitely not a good choice for page-sized allocation, we'd want a different allocator for that (or.. I dunno, actually you could use the same code, but make chunks 4096 bytes instead of 8, and jigger the field successors so the metadata gets stored elsewhere -- would be some major surgery but would work well enough).
 
x86 (our only MMU target currently) is going in a direction where any use of the MMU will virtualize the address space, making a page slab pool sufficient if we want to map X pages of anonymous memory, so no need to worry about it.

> As far as deprecating k_mem_pool_*(), I think that makes sense. But note that we have some auxilliary APIs (mbox, msgq, et. al.) which are written to assume the same k_mem_block struct that we have from mem_pool, so we can't completely deprecate it.

Yeah this would need to be addressed as part of that effort. There's a few drivers too which do this IIRC.

**Comment by npitre** [comment]
> > No. There is some metadata around memory blocks that is sure to misalign things. And assuming (1) is implemented then you'll waste every other page. For those things I think it is way way better to use a mem_slab instead.
> 
> mem_slab of page-sized blocks is only good if I just need 1 page at a time. If I want a stack of 3 pages of contiguous memory for a 12K stack buffer, a slab doesn't work (unless of course all slabs are 12K)

Would be much easier and efficient to extend the slab API to allow for requesting an arbitrary number of contiguous blocks. Fragmentation is still an issue, but you won't get wastage from alignment and heap block metadata.

**Comment by andyross** [comment]
Found some more platform build errors hidden in the logs behind the weird hang.

Just disabled the renode platform that seems to be the problem -- this is a unit test of library code and we have coverage for riscv32 via qemu anyway.

Maybe this one will work.  See the two recent commits.

**Comment by andyross** [comment]
OK, that did it.  Shippable is showing PASS now.   Unfortunately there's a new failure in "nits".  It don't know what that is.  It's not the doc check or license check or checkpatch, the link just goes to our coding style guidelines, and nothing seems to have posted to this PR a description of what went wrong...

**Comment by npitre** [comment]
I applied my optimization patches on top of this and pushed them to #24249.
Please feel free to incorporate them into your series.

File: include/kernel.h [hunk 1]
```diff
@@ -4790,6 +4790,11 @@ void k_heap_free(struct k_heap *h, void *mem);
  * If the pool is to be accessed outside the module where it is defined, it
  * can be declared via
  *
+ * @note Note: when CONFIG_MEM_POOL_HEAP_BACKEND is enabled, the
```
Comment (nashif) [inline include/kernel.h:4793]: No need for the "Note:" here

File: tests/kernel/mem_pool/mem_pool/src/main.c [hunk 2]
```diff
@@ -282,6 +317,11 @@ void alternate_task(void)
  * amount of usable space, due to the hidden block descriptor info the
  * kernel adds at the start of any block allocated from this memory pool.)
  *
+ * NOTE: when CONFIG_MEM_POOL_HEAP_BACKEND is in use, the splitting
```
Comment (nashif) [inline tests/kernel/mem_pool/mem_pool/src/main.c:320]: \@note ?

**Comment by nashif** [comment]
@andyross  Can you add a section in the documentation about all of this?

**Comment by andrewboie** [comment]
> Would be much easier and efficient to extend the slab API to allow for requesting an arbitrary number of contiguous blocks. Fragmentation is still an issue, but you won't get wastage from alignment and heap block metadata.

We're getting off on a tangent here, but we're not adding that to slabs. That's not what slabs are used for. Slabs are pools of equally sized kernel objects, our implementation uses a simple linked list stored within unused blocks. The typical use-case is each slab member to be some kernel object of a particular type.

Once you start adding support for contiguous regions this instead turns into a general purpose allocator with a minimum block size of a page, and the people who use slabs for their intended design purpose will not appreciate the footprint and overhead this entails.

**Comment by npitre** [comment]
On Thu, 9 Apr 2020, Andrew Boie wrote:

> We're getting off on a tangent here, but we're not adding that to slabs. That's not what slabs are used for. Slabs are pools of equally sized kernel objects, our implementation uses a simple linked list stored within unused blocks. The typical use-case is each slab member to be some kernel object of a particular type.

I know. It just seemed a better fit if you need to frequently allocate 3 
contiguous pages. But I didn't think it through. Sorry for the noise.

**Comment by andyross** [comment]
Rebase, fix up the "@note Note" bit.

**Comment by andyross** [comment]
Any chance to get this in before it needs another rebase?

**Pull request merged**